### PR TITLE
[core][autoscaler][v2] Reconciler - launching new instances [6/x]

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?

This PR:
- Adds the reconciling logic to compute instances to launch from QUEUED to REQUESTED 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/_private/constants.py [hunk 1]
```diff
@@ -50,6 +50,13 @@ def env_integer(key, default):
     "AUTOSCALER_MAX_CONCURRENT_LAUNCHES", 10
 )
 
+# Default upscaling speed for the autoscaler. This specified how many nodes
+# to request at a time, where the desired number to upscale is
+#   upscaling_speed * current_num_nodes.
+# e.g. 1.0 means to request enough nodes to double
+# the cluster size in each round of requests.
+DEFAULT_UPSCALING_SPEED = 0.0
```
Comment (jjyao) [inline python/ray/autoscaler/_private/constants.py:58]: 0 means we never add new nodes?

File: python/ray/autoscaler/_private/constants.py [hunk 2]
```diff
@@ -50,6 +50,13 @@ def env_integer(key, default):
     "AUTOSCALER_MAX_CONCURRENT_LAUNCHES", 10
 )
 
+# Default upscaling speed for the autoscaler. This specified how many nodes
```
Comment (jjyao) [inline python/ray/autoscaler/_private/constants.py:53]:
```suggestion
# Default upscaling speed for the autoscaler. This specifies how many nodes
```

File: python/ray/autoscaler/v2/instance_manager/reconciler.py [hunk 3]
```diff
@@ -601,3 +621,129 @@ def _reconciled_im_status_from_ray_status(
...
+        for instance in instances:
+            if instance.status == IMInstance.QUEUED:
+                queued_instances.append(instance)
+            elif instance.status == IMInstance.REQUESTED:
+                requested_instances.append(instance)
+            elif instance.cloud_instance_id:
+                allocated_instances.append(instance)
+
+        if not queued_instances:
+            # No QUEUED instances
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/reconciler.py:645]: When do we add QUEUED?

File: python/ray/autoscaler/v2/instance_manager/reconciler.py [hunk 4]
```diff
@@ -601,3 +621,129 @@ def _reconciled_im_status_from_ray_status(
...
+        )
+
+        # Transition the instances to REQUESTED for instance launcher to
+        # launch them.
+        updates = {}
+        for instance in to_launch:
+            # Reuse launch request id for any QUEUED instances that have been
+            # requested before due to retry.
+            launch_request_id = (
+                str(time.time_ns())
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/reconciler.py:663]: should we use uuid to be more unique?

File: python/ray/autoscaler/_private/constants.py (hunk 1 above)
Comment (rickyyx) [reply python/ray/autoscaler/_private/constants.py:58]: current (legacy) behaviour is always sacling up by 1 node if needed to scale up. so something like below for scaling up a specific node type:
`num_scale_up = min(1, scaling_speed * (num_allocated+num_requested))`

File: python/ray/autoscaler/v2/instance_manager/reconciler.py (hunk 3 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/reconciler.py:645]: we will add QUEUED when we scale up the cluster (after consulting the scheduler), and this process is before this.

File: python/ray/autoscaler/_private/constants.py (hunk 1 above)
Comment (jjyao) [reply python/ray/autoscaler/_private/constants.py:58]: Add this to the comment

File: python/ray/autoscaler/v2/instance_manager/reconciler.py [hunk 5]
```diff
@@ -593,3 +613,131 @@ def _reconciled_im_status_from_ray_status(
...
+
+        # Transition the instances to REQUESTED for instance launcher to
+        # launch them.
+        updates = {}
+        for instance_type, instances in to_launch.items():
+            for instance in instances:
+                # Reuse launch request id for any QUEUED instances that have been
+                # requested before due to retry.
+                launch_request_id = (
+                    str(time.time_ns())
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/reconciler.py:656]: uuid

**Review by jjyao — APPROVED** [review]

**Pull request merged**